### PR TITLE
Cleanup versioning strategy

### PR DESCRIPTION
**Pull request by baylesj**
Currently, versioning is a mess. CMake and Meson have seperate build
version number storage locations, with no way of knowing you need to
have both. Plus, due to recent revisions the amalgamate script is broken
unless you build first, and may still be broken afterwards.

This PR fixes some issues with versioning, and adds comments clarifying
what has to be done when doing a release.

#988

### CODE REVIEW

**Comment by baylesj** [comment]
Failures are due to us overusing resources. I guess I'll wait?

```
Submit files (using https)
   Using HTTP submit method
   Drop site:https://my.cdash.org/submit.php?project=jsoncpp
   Submission failed: Maximum number of builds reached for jsoncpp.  Contact webmaster@cdash.org for support.
   Submit failed, waiting 5 seconds...
   Retry submission: Attempt 1 of 3
```

**Comment by baylesj** [comment]
#734

File: include/json/version.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+#define JSONCPP_VERSION_STRING "1.9.2"
+#define JSONCPP_VERSION_MAJOR 1
+#define JSONCPP_VERSION_MINOR 9
+#define JSONCPP_VERSION_PATCH 2
+#define JSONCPP_VERSION_QUALIFIER
+#define JSONCPP_VERSION_HEXA                                                   \
+  ((JSONCPP_VERSION_MAJOR << 24) | (JSONCPP_VERSION_MINOR << 16) |             \
+   (JSONCPP_VERSION_PATCH << 8))
+
+#ifdef JSONCPP_USING_SECURE_MEMORY
```
Comment (BillyDonahue) [inline include/json/version.h]: What's this stuff about `JSONCPP_USING_SECURE_MEMORY` ?
It has nothing to do with version.
Actually JSONCPP_VERSION_HEXA and JSONCPP_VERSION_STRING are derivatives and should be elsewhere. I would like it if version.h were just the number tuple and that's about it.

File: CMakeLists.txt [hunk 2]
```diff
@@ -63,11 +63,18 @@ if(NOT DEFINED CMAKE_BUILD_TYPE)
...
 project(JSONCPP
-        VERSION 1.9.0 # <major>[.<minor>[.<patch>[.<tweak>]]]
+        # Note: version must be updated in three places when doing a release. This
+        # annoying process ensures that amalgamate, CMake, and meson all report the
+        # correct version.
+        # 1. /meson.build
+        # 2. /include/json/version.h
+        # 3. /CMakeLists.txt
+        # IMPORTANT: also update the SOVERSION!!
+        VERSION 1.9.2 # <major>[.<minor>[.<patch>[.<tweak>]]]
```
Comment (BillyDonahue) [inline CMakeLists.txt]: CMake's file command can read a variable from a file.

    file (STRINGS "version.txt" JSONCPP_VERSION)
    file (STRINGS "soversion.txt" JSONCPP_SOVERSION)

I don't know how to do it in meson.
I guess a worst-case would be invoking cat.

    r = run_command('cat','version.txt')
    JSONCPP_VERSION=r.stdout()

    r = run_command('cat','soversion.txt')
    JSONCPP_SOVERSION=r.stdout()

I don't know meson. It feels like we should be able to do this though!

Comment (baylesj) [reply CMakeLists.txt]: I'll spend some time this week looking into that, I definitely agree with you.

**Pull request merged**